프로젝트 관리 앱 [STEP 2] yeton

yeeton37 opened the pull request:
안녕하세요 찰리!!🐻 @kcharliek 

이번에 제출하는 STEP 2-1은  `UIKit + MVVM + 클린아키텍처 구조`를 적용해보았습니다.
MVVM 구조에 대해 처음 공부하다다 보니, 틀만 잡고 내부 구현은 제대로 하지 못했다는 느낌이 듭니다. 
그래서 혼자 고민하기보다 피드백을 빠르게 받고자 PR을 제출합니다!



### [UIKIt + MVVM]

- model, view, viewModel 구현 및 클린아키텍처 구조 잡아놓기
- 바인딩 구현 및 클린아키텍처 적용

### [RxSwift + MVVM]

프로젝트가 끝난 뒤에 개인적으로 공부하여 적용해볼 예정입니다!



## 🔥 고민했던 점

1️⃣  MVVM 패턴의 구조

MVVM 구조로 구현할 때, View와 Service 사이에서 데이터 변화를 관측하고, 스스로 UI를 변경할 수 있도록 하기 위한 방법에 대해 생각해보았습니다.

찰리가 말씀해주셨던 `Closure`를 이용한 방법으로 바인딩을 하고자했습니다. `Closure`를 이용한 방식은 Service내에 closure를 정의하고, 호출이 필요한 곳에 위치시켜둔 후 ViewController에서 해당 클로저를 구현하여 작동하는 방식입니다. 

하지만 클로저에 대한 이해가 부족하여  delegate를 통해 전달해주는 방식을 이용했습니다. 그래서 MVVM 패턴을 제대로 적용하지 못한 것 같다는 생각이 듭니다.

<br/>


2️⃣  Clean Architecture

MVVM에 Clean Architecture 구조를 적용하려고 했습니다.
저는 클린아키텍처를 아래와 같은 구조로 이해하였고, 이 구조를 그대로 프로젝트에 적용하면 될 것이라고 생각하여 폴더를 동일하게 나누어 주었습니다.

- Domain
    - Entity
        - 네트워킹을 통해 혹은 local DB/ remote DB를 통해 받아오는 데이터
    - Use cases
        - Entity -> Model 타입으로 변경 후 반환
- Presentation
    - Service
        - 다른 곳에서는 ViewModel이라고 칭하는 곳도 많은데, 뷰컨트롤러에 대한 로직이 담겨있는 곳이므로 Service라고 네이밍했습니다.
        - ViewModel -> Model, Model -> ViewModel로 변경
    - ViewController
        - viewModel 가져와 사용
- Data
    - Repository
        - Entity를 클로저로 반환
- 기타
    - Model
    - ViewModel
        - model을 뷰에 보이는 형식으로 바꿈(ex. Date -> String)

<br/>

3️⃣  화면마다 하나의 Service 생성

각 뷰컨트롤러마다 `Service` 타입을 만들어주어 사용하고 싶었으나 메인 뷰컨트롤러가 아닌 다른 뷰컨트롤러에서는 크게 데이터 전달이 필요한 로직이 많이 요구되지 않아 하나의 `Service`만 생성해주었습니다.

<br/>

4️⃣ 뷰모델 내부의 input과 output 구현

찰리와 디스코드에서 나눈 이야기를 토대로 Service 내부에 `Input`과 `Output` 타입을 구현해주었습니다. 

`Input`은 ViewController에서 Service로 이벤트가 발생했음을 알리고, `Output`은 데이터를 화면에 보이는 타입으로 변경하여 Service에서 ViewController로  뷰모델 타입의 데이터를 전달해주는 역할을 한다고 이해했습니다.

따라서 ViewController에서 어떤 이벤트가 발생했을 때마다 클로저를 통해 Service의 input에 구현된 클로저를 호출해주었습니다. 

찰리가 말씀하신 데이터 전달의 흐름에 맞게 잘 구현한 것인지 궁금합니다! 👀👀

<br/>

5️⃣ Repository 및 Usecase 추상화

추후에 프로젝트 매니저2를 진행할 때 `Remote`의 `Repository`도 구현하게 될텐데, 이를` RepositoryProtocol` 타입을 채택하여 구현하면 나중에 갈아끼우기만 해도 코드가 작동할 수 있도록 수정해주었습니다.

-> ***찰리와 이야기를 통해 리팩토링 진행했습니다!***

<br/>

6️⃣ 데이터의 전달 방식 고민

- 메인 뷰컨트롤러 -> 수정 뷰컨트롤러 or 생성 뷰컨트롤러: 의존성 주입을 통해 데이터를 바로 전달
- 수정 뷰컨트롤러 or 생성 뷰컨트롤러 -> 메인 뷰컨트롤러: `delegate` 패턴을 통해 데이터를 전달하여 의존 관계를 역전

-> ***이 또한 찰리와 이야기를 통해 리팩토링 진행했습니다! 구현하려고 보니,, 제가 아무생각 없이 이미 delegate 패턴을 쓰고 있었더군요ㅠㅠ 반성중입니다,,***




---

<br/>

이상입니다 많이 신경써주셔서 정말 감사합니다 찰리!
잘부탁드립니다!👍

yeeton37 committed:
feat: ModelDTO 생성 및 ProjectState를 나타내는 enum 생성

yeeton37 committed:
feat: TableViewCell 생성

yeeton37 committed:
feat: TaskRepository 프로토콜 및 TaskRepository 구현

yeeton37 committed:
feat: TaskUseCase 프로토콜 및 TaskUseCase 구현

yeeton37 committed:
feat: TaskViewModel 프로토콜 및 TaskViewModel 구현

- swiftLint 수정

yeeton37 committed:
feat: TaskTableView 및 TaskHeaderView 구현

yeeton37 committed:
feat: Main과 Edit 각 화면 ViewController 생성

- 그에 따른 MainTaskViewControllerDelegate 생성
- EF_Diary 폰트 적용

yeeton37 committed:
file: 폴더 구조 변경

yeeton37 committed:
feat: 셀 클릭시 나타나는 VC의 title, date, body 설정

kcharliek commented on ProjectManager/ProjectManager/Domain/UseCases/TaskUseCase.swift:
```diff
@@ -0,0 +1,28 @@
+import Foundation
+
+class TaskUseCase: TaskUseCaseProtocol {
+    private var repository = TaskRepository()
```
상세 타입이 아니라 Protocol로 참조 해주세요.

kcharliek replied:
'의존성 주입' 이라는 키워드로 구글링 해보세요.

yeeton37 replied:
프로토콜로 참조하여 TaskRepository를 추상화해주는 것으로 리팩토링 완료했습니다!

kcharliek commented:
이야기가 길어져서 음성으로 리뷰 진행 중입니다.

yeeton37 committed:
refactor: entity 타입 수정 및 model 생성

yeeton37 committed:
refactor: Date와 String 타입 확장

yeeton37 committed:
refactor: LocalTaskRepository 추상화 및 프로토콜 수정

yeeton37 committed:
refactor: TaskUseCase 추상화 및 프로토콜 수정

yeeton37 committed:
refactor: MainTaskService 추상화 및 프로토콜 생성

yeeton37 committed:
feat: 모델 역할을 하는 TaskViewModel 생성

- index를 가져오기 위해 Equatable 채택

yeeton37 committed:
refactor: 전반적인 네이밍 수정, 기존 코드 삭제, swiftLint 수정

yeeton37 committed:
refactor: CreateTaskViewControllerDelegate 생성

yeeton37 committed:
refactor: EditTaskViewControllerDelegate 생성

- 기존에 있던 MainTaskViewControllerDelegate 삭제

yeeton37 committed:
refactor: EditTaskViewController 수정

- 기존 entity 타입의 변수를 viewModel 타입으로 변경
- 오토레이아웃의 컨벤션 수정

yeeton37 committed:
refactor: MainTaskViewController 전반적인 수정

- mainTaskService의 output 클로저를 뷰컨트롤러 내부에 구현
- configureLongPressGestureRecognizer 메서드 내부 로직 수정

yeeton37 committed:
style: 폴더 구조 정리

yeeton37 committed:
docs: README.md 수정

yeeton37 committed:
docs: README.md 수정

kcharliek commented on ProjectManager/ProjectManager/Domain/Entities/Enum/TaskState.swift:
```diff
@@ -0,0 +1,36 @@
+
+import Foundation
+
+enum TaskState {
```
Enum 분리 굿입니다 :)

yeeton37 replied:
감사합니다!!

kcharliek commented on ProjectManager/ProjectManager/Domain/UseCases/TaskUseCase.swift:
```diff
@@ -0,0 +1,47 @@
+import Foundation
+
+struct TaskUseCase: TaskUseCaseProtocol {
+    var repository: TaskRepositoryProtocol
```
private 처리 해주세요~

yeeton37 replied:
private 처리 해주었습니다!

kcharliek commented on ProjectManager/ProjectManager/Presentation/View/MainTaskViewController.swift:
```diff
@@ -0,0 +1,276 @@
+            return UITableViewHeaderFooterView()
+        }
+        
+        if let state = (tableView as? TaskTableView)?.projectState,
```
저라면 tableView에서 data를 직접 꺼내도록 하지 않았을 것 같아요.
View는 View로만 두는 것이 가장 아름답습니다. 
View 스스로 상태를 담고 있는 모양새는 어쩌면 비즈니스 로직을 들고있는 것과 같습니다.

저라면

func state(of: tableView) {
if tableView == taskFinishedTableView { return .someState }
else {...} 
}

이런 식으로 분기했을 것 같네요 .

yeeton37 replied:
테이블뷰가 현재 자신의 상태를 알고 있는 형태는 비즈니스 로직을 들고있는 것과 같군요! 처음 알았습니다

### View Controller

```swift
private func state(of tableView: UITableView) -> TaskState {
   if tableView == taskStandbyTableView {
       return .todo
   } else if tableView == taskInProgressTableView {
       return .doing
   } else {
      return .done
   }
}
```

```swift
 let state = state(of: tableView)
```

찰리의 말씀대로 위와 같이 state 메서드를 만들어서 사용해주는 것으로 변경했습니다!

kcharliek commented on ProjectManager/ProjectManager/Presentation/View/MainTaskViewController.swift:
```diff
@@ -0,0 +1,276 @@
+                return
+            }
+            
+            self.mainTaskService.input.deleteButtonDidTap(indexPath.row, viewModel[indexPath.row].state)
```
비즈니스 로직이네요.

이쪽은 ViewModel의 input쪽으로 이벤트를 보내주세요 .

kcharliek replied:
viewModel.input.deleteActionRaised 등.

kcharliek commented on ProjectManager/ProjectManager/Presentation/View/MainTaskViewController.swift:
```diff
@@ -0,0 +1,276 @@
+            }
+            
+            self.mainTaskService.input.deleteButtonDidTap(indexPath.row, viewModel[indexPath.row].state)
+            self.reloadTableView()
```
reloadTableView 를 하는 것도 Service가 판단하도록 흐름을 변경해주세요 .

yeeton37 replied:
reloadTableView 하는 부분 모두 Service에서 판단하여 실행하도록 수정해주었습니다!

yeeton37 replied:
### View Controller (호출)

```swift
self.mainTaskService.input.deleteButtonDidTap(
indexPath.row, 
viewModel[indexPath.row].state
)
```

### Service

```swift
 func deleteButtonDidTap(
        index: Int,
        state: TaskState
    ) {
        let task = taskUseCase.fetch().filter { $0.state == state }[index]
        
        taskUseCase.delete(task: task)
        output.deleteCell(index, state)
        output.reloadTableView()
    }
```

kcharliek commented on ProjectManager/ProjectManager/Presentation/View/MainTaskViewController.swift:
```diff
@@ -0,0 +1,276 @@
+        mainTaskService.taskUseCase.update(task: newTask)
+        fetchedViewModels?.remove(at: index)
+        fetchedViewModels?.insert(changedViewModel, at: index)
+        reloadTableView()
```
이쪽 비즈니스 로직을 직접 구현하고있네요 . 

흐름을 Service쪽으로 넘겨주세요 .

yeeton37 replied:
엇... 놓친 부분이네요! 

### Service
```swift
 func editDoneButtonDidTap(viewModel: TaskViewModel) {
    let task = Task.init(viewModel: viewModel)
        
    taskUseCase.update(task: task)
}
```
### ViewController (호출)
```swift
 mainTaskService.input.editDoneButtonDidTap(changedViewModel)
```

위와 같이 수정해 비즈니스 로직을 viewModel로 넘겨주었습니다!

kcharliek reviewed (approved):
고생하셨습니다 :)
